internal/lib/sandbox/builder.go

```diff
@@ -0,0 +1,379 @@
+package sandbox
+
+import (
+	"errors"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/containers/storage/pkg/idtools"
+	"github.com/containers/storage/pkg/stringid"
+	"github.com/opencontainers/runtime-tools/generate"
+	types "k8s.io/cri-api/pkg/apis/runtime/v1"
+
+	"github.com/cri-o/cri-o/internal/factory/container"
+	"github.com/cri-o/cri-o/internal/hostport"
+	"github.com/cri-o/cri-o/internal/memorystore"
+	"github.com/cri-o/cri-o/internal/oci"
+	"github.com/cri-o/cri-o/internal/storage"
+	libconfig "github.com/cri-o/cri-o/pkg/config"
+)
+
+// Builder is the interface for managing pod sandboxes.
+type Builder interface {
+	// SetConfig sets the sandbox configuration and validates it
+	SetConfig(*types.PodSandboxConfig) error
+
+	// SetNameAndID sets the sandbox name and ID
+	SetNameAndID() error
+
+	// Config returns the sandbox configuration
+	Config() *types.PodSandboxConfig
+
+	// ID returns the id of the pod sandbox
+	ID() string
+
+	// Name returns the id of the pod sandbox
+	Name() string
+
+	// InitInfraContainer initializes the sandbox's infra container
+	InitInfraContainer(*libconfig.Config, *storage.ContainerInfo, *idtools.IDMappings) error
+
+	// Spec returns the infra container's generator.
+	// Must be called after InitInfraContainer.
+	Spec() *generate.Generator
+
+	// ResolvPath returns the sandbox's resolvPath.
+	ResolvPath() string
+
+	// SetDNSConfig sets the dns configs.
+	SetDNSConfig(*types.DNSConfig)
+
+	// SetCriSandbox sets and creates CriSandbox.
+	SetCriSandbox(string, time.Time, map[string]string, map[string]string, *types.PodSandboxMetadata)
```

Comment:
nit: can this be `SetCRISandbox`

Reply:
also I feel like we could do this step at sandbox generation step, and fail if the respective fields aren't defined. We pass all of these things in through setters beforehand right? what if a caller messes up and sets them differently? I think we should inherit as much as possible to reduce code

Reply:
So now we will use the same createdAt to set the time here and hence removed it from params but now it does require to set created at before hand else it will throw error